### PR TITLE
Increasing test coverage rclcpp_lifecycle

### PR DESCRIPTION
**Pull request by brawner**
As with #1043 and #1044, this PR increases the test coverage for rclcpp_lifecycle. The test coverage increases from 61% to 100% of the `include` files and 87% of the `src` files (88% combined)

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/9864/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5480/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8024/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9720/)
* Coverage [Build Status](https://ci.ros2.org/job/ci_linux_coverage/78/)

Depends on #1043, #1044
New commit: https://github.com/ros2/rclcpp/commit/7441e1da2454b1e82aa1bbf9a095df766f129382

### CODE REVIEW

File: rclcpp_action/CMakeLists.txt [hunk 1]
```diff
@@ -98,6 +98,16 @@ if(BUILD_TESTING)
       ${PROJECT_NAME}
     )
   endif()
+
```
Comment (ivanpauno) [inline rclcpp_action/CMakeLists.txt:101]: this PR also adds a lot of tests for `rclcpp_action`, were those added intentionally here?
Seems that the same tests were added in https://github.com/ros2/rclcpp/pull/1043

File: rclcpp_components/CMakeLists.txt [hunk 2]
```diff
@@ -73,12 +73,21 @@ if(BUILD_TESTING)
   set(components "${components}test_rclcpp_components::TestComponentFoo;$<TARGET_FILE:test_component>\n")
   set(components "${components}test_rclcpp_components::TestComponentBar;$<TARGET_FILE:test_component>\n")
   set(components "${components}test_rclcpp_components::TestComponentNoNode;$<TARGET_FILE:test_component>\n")
+  set(components "${components}test_rclcpp_components::TestComponentThrows;$<TARGET_FILE:test_component>\n")
+
+  # Used to catch invalid resource entry
+  set(invalid_components "test_rclcpp_components::TestComponentFoo;;$<TARGET_FILE:test_component>\n")
```
Comment (ivanpauno) [inline rclcpp_components/CMakeLists.txt:79]: Same about adding tests to `rclcpp_components`, sounds like a duplicate of https://github.com/ros2/rclcpp/pull/1044.

**Comment by brawner** [comment]
@ivanpauno Because there was a lot of files added in these PRs, I broke them out into three different ones stacked on top of one another. There is no reason why they should be dependent on one another. I'll clean that up.

**Comment by brawner** [comment]
Rebased on master and removed dependency from #1043 and #1044.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10564/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6017/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8604/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10446/)

File: rclcpp_lifecycle/test/test_lifecycle_service_client.cpp [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+  {
+    rclcpp::init(0, nullptr);
+    lifecycle_node_ = std::make_shared<EmptyLifecycleNode>();
+    lifecycle_client_ = std::make_shared<LifecycleServiceClient>("client");
+    spinner_ = std::thread(&TestLifecycleServiceClient::spin, this);
+  }
+
+  void TearDown() override
+  {
+    rclcpp::shutdown();
```
Comment (dirk-thomas) [inline rclcpp_lifecycle/test/test_lifecycle_service_client.cpp:220]: Calling `shutdown` from a different thread causes a race which was making this test fail in the nightly jobs since it has been added. See #1204 for the proposed fix.
